Governance feature implementation

chetan-zilliqa opened the pull request:
## Description
<!-- What is the overall goal of your pull request? -->
<!-- What is the context of your pull request? -->
<!-- What are the related issues and pull requests? -->
The Governance feature is the implementation of zip [](https://github.com/Zilliqa/ZIP/pull/34/files).All miners in Zilliqa community can vote on the proposal put forward by zilliqa team and that makes a permanent record in blockchain.

## Implementation details
1) All miners will have voting script ,through which they will cast their votes.The vote will be stored temporary in `Node` class object.
2) When the next pow submission happens, the vote will be sent as a part of pow submission message to DS nodes.
3) At DS nodes , all votes from the miners are accumulated and made as a part of DS block.
4) DS Leader will run consensus on the DS block and upon successful consensus, votes will be part of DS block.


NOTE : 
1) `scripts/governance_proposal_vote.py`, is kept only in PR for testing purpose, i will remove the script before final merge and raise a separate PR in `testnet` repo for the miner script.
2) LOGS and TODO's mentioned in the code is just for debugging purpose.I will remove them before doing final merge.


`GetDsBlock` api response
`{
  "id": "1",
  "jsonrpc": "2.0",
  "result": {
    "Governance": [
      {
        "ProposalId": 12345,
        "ShardVotes": [
          {
            "VoteCount": 5,
            "VoteValue": 1
          },
          {
            "VoteCount": 7,
            "VoteValue": 2
          },
          {
            "VoteCount": 3,
            "VoteValue": 3
          }
        ]
      }
    ],
    "header": {
      "BlockNum": "3",
      "Difficulty": 3,
      "DifficultyDS": 5,
      "GasPrice": "1000000000",
      "LeaderPubKey": "0x02BE864EABAAFC97CF7B794B7775A45B3ADD009C4BEBB58462EEA4E36B6F4F0236",
      "PoWWinners": [],
      "PrevHash": "2ac44cfb038f961a7a1eb6a161b16b5aa0a4510df2cc48eda43cccea343d6811",
      "Timestamp": "1598410460175886"
    },
    "signature": "E51A8C81DBE0B441D6F78190121300AB5D47C6465C8B0F294FEBD8CED9BADAE78E2359118F6CC224BF99C57AED3BAACCA79A1871449762969FF4861E4FC74098"
  }
}`
## Backward Compatibility
<!-- For breaking changes, code must be protected by UPGRADE_TARGET -->
- [ ] This is not a breaking change
- [ ] This is a breaking change

## Review Suggestion
<!-- How should the reviewers get started on reviewing your pull request-->
<!-- How can the reviewers verify the pull request is working as expected -->

## Status

### Implementation
<!-- Add more TODOs before "ready for review", if any  -->
- [ ] **ready for review**

### Integration Test (Core Team)
<!-- This is for core team only, ignore this if you are a community contributor -->

<!-- append the commit digest to inform the others of the versions you have tested -->
- [x] local machine test
<!-- - [ ] local machine test (commit: bbbbbbbb) -->
<!-- - [ ] local machine test (commit: cccccccc) -->
<!-- - [ ] local machine test (commit: dddddddd) -->
- [x] small-scale cloud test
<!-- - [ ] small-scale cloud test (commit: bbbbbbbb) -->
<!-- - [ ] small-scale cloud test (commit: cccccccc) -->
<!-- - [ ] small-scale cloud test (commit: dddddddd) -->

ansnunez commented on src/libMessage/Messenger.cpp:
```diff
@@ -1390,6 +1403,34 @@ void DSBlockHeaderToProtobuf(const DSBlockHeader& dsBlockHeader,
+      protoproposal = protoDSBlockHeader.add_proposals();
+      protoproposal->set_proposalid(govProposal.first);
+      for (const auto& vote : govProposal.second.first) {
+        LOG_GENERAL(INFO, "Governance: serialization DS nodes govProposalId: "
```
Avoid long log messages with a lot of colons used for both labeling ("Governance:") and separating key-value pairs (quite hard to read).

Also, aim for consistency in log formats (for example, "govVoteValue:" has no trailing space but "voteCount: " has).

You might want to go with something like this instead (note: no spaces around `=`):
```
[Gov][DS] Proposal=<ID> Value=<value> Count=<count>
[Gov][Shards] Proposal=<ID> Value=<value> Count=<count>
```

chetan-zilliqa replied:
yeah. I changed the log lines as suggested,
But this log , i will remove later on.I added them to verify if serialization and de-serialization works fine.

ansnunez commented on src/libMessage/Messenger.cpp:
```diff
@@ -3885,6 +3962,19 @@ bool Messenger::SetDSPoWSubmission(
+    if (!result.data().govdata().IsInitialized()) {
+      LOG_GENERAL(WARNING,
+                  "DSPoWSubmission governance data initialization failed");
+    }
```
Do you want to return false here? Just checking

chetan-zilliqa replied:
Intentionally not returning from here.Otherwise it will impact normal behavior if theirs some problem in governance message serialization.

ansnunez commented on src/libNode/Node.cpp:
```diff
@@ -3224,4 +3224,12 @@ void Node::CleanLocalRawStores() {
+}
+void Node ::storeVoteUntilPow(const std::string& proposalId,
+                              const std::string& vote) {
+  m_govProposal = make_pair(static_cast<uint32_t>(std::stoul(proposalId)),
```
Can this potentially throw an uncaught exception?

chetan-zilliqa replied:
yeah.correct. Fixing it. `stoul` can throw exception.Thanks.

ansnunez commented on src/libData/MiningData/DSPowSolution.h:
```diff
@@ -35,6 +35,8 @@ class DSPowSolution {
   std::string m_mixHash;
   uint32_t m_lookupId{};
   uint128_t m_gasPrice;
+  std::pair<uint32_t, uint32_t>
```
Also, should we make this an atomic type? In case for example `storeVoteUntilPow` is called at the same time this value is cleared out (or other read/write race condition scenarios)

chetan-zilliqa replied:
Created.
`std::pair<std::atomic<uint32_t>, std::atomic<uint32_t>> m_govProposal;
  std::atomic<uint32_t> m_govMaxVoteAttempt{};`

chetan-zilliqa commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -974,6 +974,40 @@ bool DirectoryService::RunConsensusOnDSBlockWhenDSPrimary() {
+    if (powSolIter != allDSPoWs.end()) {
+      uint32_t proposalId = powSolIter->second.m_govProposal.first;
+      uint32_t voteValue = powSolIter->second.m_govProposal.second;
+      LOG_GENERAL(INFO, "Governance: DS votes count : proposalId: "
```
Actually LOG_GENERAL added by me throughout the PR is only kept for debugging purpose at present.I thought to remove  them altogether once PR is ready to be merged.

chetan-zilliqa commented on src/libData/MiningData/DSPowSolution.cpp:
```diff
@@ -99,7 +113,10 @@ bool DSPowSolution::operator==(const DSPowSolution& sol) const {
           (m_mixHash == sol.m_mixHash) && (m_lookupId == sol.m_lookupId) &&
-          (m_gasPrice == sol.m_gasPrice) && (m_signature == sol.m_signature));
+          (m_gasPrice == sol.m_gasPrice) &&
+          (m_govProposal.first == sol.m_govProposal.first) &&
```
yes.No need to compare individual elements.changed code

chetan-zilliqa commented on src/libData/MiningData/DSPowSolution.h:
```diff
@@ -84,6 +87,12 @@ class DSPowSolution {
   const uint128_t& GetGasPrice() const;
 
+  /// Returns proposal id
+  const uint32_t& GetGovProposalId() const;
```
done

chetan-zilliqa commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -974,6 +974,40 @@ bool DirectoryService::RunConsensusOnDSBlockWhenDSPrimary() {
+  for (const auto& dsnode : powDSWinners) {
+    const auto& powSolIter = allDSPoWs.find(dsnode.first);
+    if (powSolIter != allDSPoWs.end()) {
+      uint32_t proposalId = powSolIter->second.m_govProposal.first;
```
done

chetan-zilliqa commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -974,6 +974,40 @@ bool DirectoryService::RunConsensusOnDSBlockWhenDSPrimary() {
+    const auto& powSolIter = allDSPoWs.find(dsnode.first);
+    if (powSolIter != allDSPoWs.end()) {
+      uint32_t proposalId = powSolIter->second.m_govProposal.first;
+      uint32_t voteValue = powSolIter->second.m_govProposal.second;
```
done

chetan-zilliqa commented on src/libDirectoryService/DirectoryService.h:
```diff
@@ -48,23 +48,33 @@ struct PoWSolution {
+        m_govProposal(gvp) {}
   bool operator==(const PoWSolution& rhs) const {
-    return std::tie(m_nonce, m_result, m_mixhash, m_lookupId, m_gasPrice) ==
+    return std::tie(m_nonce, m_result, m_mixhash, m_lookupId, m_gasPrice,
```
done

chetan-zilliqa commented on src/libDirectoryService/PoWProcessing.cpp:
```diff
@@ -283,9 +288,11 @@ bool DirectoryService::VerifyPoWSubmission(const DSPowSolution& sol) {
+  LOG_GENERAL(INFO, "Key          = " << submitterPubKey);
+  LOG_GENERAL(INFO, "Peer         = " << submitterPeer);
+  LOG_GENERAL(INFO, "Diff         = " << to_string(difficultyLevel));
+  LOG_GENERAL(INFO, "GovProposalId   = " << to_string(govProposalId));
```
done

chetan-zilliqa commented:
I have changed the code further to accommodate below behavior mentioned in the zip.
"Finally, the implementation should be such that a node that fails to get sharded should be able to re-attempt casting its vote in the next DS epoch, without having to configure the node once again"
For that, introduced new variable `m_govMaxVoteAttempt` which will be set by the miner script. The variables tracks how many POW submission should be allowed to cast votes.Additionally we can keep in constant.xml for the default value if needed.
On doing POW submission , node will reduce `m_govMaxVoteAttempt` by 1.If it becomes 0, then resetting governance proposal id and vote to 0.
If the node  becomes part of shard and DS Member.Node receives `ProcessVCDSBlocksMessage`.I am resetting the proposal id and vote to 0, so that in next POW submission, the proposal id and votes is zero and  are not counted.

sandipbhoir commented on src/libServer/StatusServer.cpp:
```diff
@@ -481,4 +485,32 @@ string StatusServer::GetValidateDB() {
+    return false;
+  }
+  try {
+    if (!m_mediator.m_node->StoreVoteUntilPow(proposalId, voteValue,
```
With current setup, `m_govProposalInfo` can be overwritten on receiving new proposal through api and can potentially result in undefined behavior.

ansnunez replied:
For now we can allow miner to change the vote settings anytime. As long as `m_mutexGovProposal` is acquired wherever `m_govProposalInfo` is used, then that should avoid any race condition and undefined state.

sandipbhoir replied:
Yes.. I agree. I should have said unexpected behaviour instead. Anyways if miners are made aware of this restriction on one proposal voting each ds epoch, it should be fine.
Later we can consider adding this restriction at code level or remove this restriction itself to allow multiple proposal voting in same ds epoch.

ansnunez commented on src/libNode/PoWProcessing.cpp:
```diff
@@ -281,6 +281,40 @@ bool Node::StartPoW(const uint64_t& block_num, uint8_t ds_difficulty,
   return true;
 }
 
+bool Node::CheckIfGovProposalActive() {
```
> Looks like only current setup will work with only one proposal being active in our network. @chetan-zilliqa @ansnunez
> `m_govProposalInfo` is tied to node which track only one proposal until its lifetime.
> 
> Is this expected behaviour?

On node side yes we support one proposal vote only inside PoW submission. If a miner wants to vote on another proposal it would have to be on a different PoW submission by calling `SetVoteInPow` again.

chetan-zilliqa commented on src/libMessage/Messenger.cpp:
```diff
@@ -1459,6 +1490,25 @@ bool ProtobufToDSBlockHeader(
+    std::map<uint32_t, uint32_t> shardVotes;
+    for (const auto& protovote : protoProposal.dsvotes()) {
+      dsVotes[protovote.value()] = protovote.count();
+      LOG_GENERAL(INFO, "Gov DS Vote=" << protovote.value()
```
@sandipbhoir , proposal id is logged at the top for loop at  line 1491.
One proposal will have multiple votes and vote count.so logging proposal id first and all the vote values and vote counts against that proposal.
